File: src/client/java/me/a0g/api/WynnItem.java
package me.a0g.api;

import com.google.gson.JsonObject;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;

/**
 * Slim-предмет с нашего бэкенда: только displayName, tier и icon.
 */
public class WynnItem {
    private final String name;
    private final String tier;
    private final JsonObject icon;
    private final boolean isCustom;

    public WynnItem(JsonObject json) {
        this(json, false);
    }

    public WynnItem(JsonObject json, boolean isCustom) {
        this.name = json.has("displayName") ? json.get("displayName").getAsString() : "";
        this.tier = json.has("tier") ? json.get("tier").getAsString().toLowerCase() : "normal";
        this.icon = json.has("icon") && json.get("icon").isJsonObject() ? json.getAsJsonObject("icon") : null;
        this.isCustom = isCustom;
    }

    public String getName() {
        return name;
    }

    public String getTier() {
        return tier;
    }

    public JsonObject getIcon() {
        return icon;
    }

    public boolean isCustom() {
        return isCustom;
    }

    /**
     * Цвет редкости. Бэкенд нормализует tier к одному из:
     * normal | unique | rare | legendary | mythic | fabled.
     */
    public Formatting getRarityColor() {
        if (isCustom) {
            return Formatting.GOLD;
        }
        return switch (tier) {
            case "unique" -> Formatting.YELLOW;
            case "rare" -> Formatting.LIGHT_PURPLE;
            case "legendary" -> Formatting.AQUA;
            case "fabled" -> Formatting.RED;
            case "mythic" -> Formatting.DARK_PURPLE;
            case "custom" -> Formatting.DARK_AQUA; //а как в getRarityColorValue
            default -> Formatting.GRAY;
        };
    }

    public int getRarityColorValue() {
        return switch (getRarityColor()) {
            case GOLD -> 0xFFAA00;
            case YELLOW -> 0xFFFF55;
            case LIGHT_PURPLE -> 0xFF55FF;
            case AQUA -> 0x55FFFF;
            case RED -> 0xFF5555;
            case DARK_PURPLE -> 0xAA00AA;
            case GRAY -> 0xAAAAAA;
            case DARK_AQUA -> 0x00FFFF;
            default -> 0xFFFFFF;
        };
    }

    public String getNameToChat() {
        return name == null ? "" : name;
    }

    public Text getFormattedName() {
        if (name == null || name.isEmpty()) {
            return Text.literal("");
        }
        if (isCustom) {
            return Text.literal("§7[Custom] §6" + name).formatted(Formatting.GOLD);
        }
        return Text.literal(name).formatted(getRarityColor());
    }

    public boolean matchesSearch(String query) {
        if (query == null || query.isEmpty() || name == null) {
            return true;
        }

        String lowerQuery = query.toLowerCase();
        String lowerName = name.toLowerCase();

        if (lowerName.contains(lowerQuery)) {
            return true;
        }

        String nameWithSpaces = lowerName.replace("-", " ");
        if (nameWithSpaces.contains(lowerQuery)) {
            return true;
        }

        String[] queryWords = lowerQuery.split("\\s+");
        String[] nameWords = nameWithSpaces.split("\\s+");

        for (String queryWord : queryWords) {
            boolean wordFound = false;
            for (String nameWord : nameWords) {
                if (nameWord.contains(queryWord)) {
                    wordFound = true;
                    break;
                }
            }
            if (!wordFound) return false;
        }
        return true;
    }
}

File: src/main/kotlin/me/a0g/wms/core/WynnItem.kt
package me.a0g.wms.core

import com.google.gson.JsonObject
import gg.essential.universal.ChatColor
import me.a0g.wms.Wms
import net.minecraft.client.Minecraft
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import scala.annotation.switch

class WynnItem(var json: JsonObject) {
    var name = json["name"].asString
//    var type = json.get("type").asString//["type"].asString
//    var armorType = json.get("armorType").asString//["armorType"].asString
//    var material = json.get("material").asString//["material"].asString

    fun getTier():String{
        val tier:String? = if(json.has("tier") && !json.get("tier").isJsonNull) json.get("tier").asString.lowercase() else null
        var color:String = ""
        when(tier) {
            "normal" -> color = ChatColor.GRAY.toString()
            "unique" -> color = ChatColor.YELLOW.toString()
            "rare" -> color = ChatColor.LIGHT_PURPLE.toString()
            "legendary" -> color = ChatColor.AQUA.toString()
            "fabled" -> color = ChatColor.RED.toString()
            "mythic" -> color = ChatColor.DARK_PURPLE.toString()
            "set" -> color = ChatColor.GREEN.toString()
        }

        return color
    }

    fun getNameToChat():String{
        val name = this.name
        if(name.isNullOrEmpty())
            return ""

        return name
    }

    fun getTextToRender():String{
        val color = getTier()
        val text = name;
        if(text.isNullOrEmpty()) return ""

        return "$color$text"
    }

    fun getItemToRender():ItemStack? {
       // Wms.logger.info(json)
        val type:String? = if(json.has("type") && !json.get("type").isJsonNull) json.get("type").asString else null
        val armorType:String? = if(json.has("armorType") && !json.get("armorType").isJsonNull) json.get("armorType").asString else null
        val material:String? = if(json.has("material") && !json.get("material").isJsonNull) json.get("material").asString else null

        var itemStack:ItemStack? = null
        if(material.isNullOrEmpty()){
            if(type != null && armorType != null){
                val item =  Item.getByNameOrId("minecraft:${armorType.lowercase().replace("chain".toRegex(),"chainmail")}_${type.lowercase()}")
                itemStack = item?.let { ItemStack(it) }
            }
        }
        else{
            val id = material.replace(":.+".toRegex(),"").toInt()
            val metadata = material.replace(".+:".toRegex(),"").toInt()
            //Wms.logger.info(metadata)
            val item = Item.getItemById(id)
            if(metadata <= 0){
                itemStack = ItemStack(item)
            }
            else{
                val iStack = ItemStack(item)
                val tag = NBTTagCompound()
                tag.setBoolean("Unbreakable",true)
                iStack.tagCompound = tag
                iStack.itemDamage = metadata
                itemStack = iStack
            }
        }

        if(itemStack != null)
            return itemStack

        return null
    }
}